miniapps/gslib/particles_redist.cpp

```diff
@@ -0,0 +1,208 @@
+   int npt = 1000;
+   bool visualization = true;
+   int visport = 19916;
+   char vishost[] = "localhost";
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh", "Mesh file to use");
+   args.AddOption(&npt, "-npt", "--num-particles",
+                  "Number of particles to initialize on global mesh "
+                  "bounding box.");
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.AddOption(&visport, "-p", "--send-port", "Socket for GLVis.");
+
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
+   }
+   if (rank == 0) { args.PrintOptions(cout); }
+
+   // Create mesh
+   Mesh mesh(mesh_file);
+   int space_dim = mesh.Dimension();
+
+   // Create parallel particle set
+   ParticleSet pset(MPI_COMM_WORLD, npt, space_dim);
+
+   Vector pos_min, pos_max;
+   mesh.GetBoundingBox(pos_min, pos_max);
+
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+
+   // Set particles randomly on entire mesh domain, for each rank
+   std::mt19937 gen(rank);
+   std::uniform_real_distribution<real_t> real_dist;
+   for (int i = 0; i < pset.GetNParticles(); i++)
+   {
+      Particle p = pset.GetParticleRef(i);
+      for (int d = 0; d < pset.GetDim(); d++)
+      {
+         p.Coords()[d] = pos_min[d] + real_dist(gen)*(pos_max[d] - pos_min[d]);
+      }
+   }
+
+   // Find points
+   FindPointsGSLIB finder(MPI_COMM_WORLD);
+   pmesh.EnsureNodes();
+   finder.Setup(pmesh);
+   finder.FindPoints(pset.Coords(), pset.Coords().GetOrdering());
+
+   // Remove points not in domain
+   const Array<int> rm_idxs = finder.GetPointsNotFoundIndices();
+   pset.RemoveParticles(rm_idxs);
+
+   int num_removed = rm_idxs.Size();
+   MPI_Allreduce(MPI_IN_PLACE, &num_removed, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+   if (rank == 0)
+   {
+      mfem::out << endl << "Removed " << num_removed <<
+                " particles that were not within the mesh" << endl << endl;
+   }
+
+   // Remove ranks from finder proc list corresponding to removed particles
+   Array<unsigned int> ranks = finder.GetProc();
+   ranks.DeleteAt(rm_idxs);
+   if (rank == 0)
+   {
+      mfem::out << "Pre-Redistribute:" << endl;
+   }
+
+   PrintOnOffRankCounts(ranks);
+
+   // Particle size for visualization
+   real_t psize = Distance(pos_min, pos_max)*2e-3;
+   if (visualization)
+   {
+      socketstream sock;
+      Vector rank_vector(pset.GetNParticles());
+      rank_vector = rank;
+      VisualizeParticles(sock, vishost, visport,
+                         pset, rank_vector, psize,
+                         "Particle Owning Rank (Pre-Redistribute)",
+                         0, 0, 400, 400, pset.GetDim() == 2 ? "Rjbcb" : "cb");
+   }
+
+
+   // Redistribute
+   pset.Redistribute(ranks);
+
+   // Find again
+   finder.FindPoints(pset.Coords(), pset.Coords().GetOrdering());
```

Comment:
This could be applied only on the redistributed particles 🤔 , but we may leave it for another day...